### PR TITLE
must-gather: use the knit package

### PR DESCRIPTION
**Pull request by fromanirh**
We add the knit tool to the must-gather image.
The tool is already useful for troubleshooting; the next step
is to proper integration with must-gather once we fill the gaps
in `knit`.

Signed-off-by: Francesco Romani <fromani@redhat.com>

### CODE REVIEW

**Comment by fromanirh** [comment]
ci's broken today

File: Makefile [hunk 1]
```diff
@@ -306,7 +316,7 @@ generate-release-tags:
 
 # Produce CRDs that work back to Kubernetes 1.11 (no version conversion)
 CRD_OPTIONS ?= "crd:crdVersions=v1"
-CONTROLLER_GEN = "$(TOOLS_BIN_DIR)/controller-gen"
+CONTROLLER_GEN = "$(TOOLS_BIN_DIR)/github.com/openshift-kni/debug-tools/cmd/knit/main.gocontroller-gen"
```
Comment (marcel-apf) [inline Makefile:319]: I don't "get" this change. Can you please explain?

Comment (fromanirh) [reply Makefile:319]: editing mistake, need to be removed

**Comment by fromanirh** [comment]
actually we can do better than this with a trivial refactoring of knit, let me work on a couple PRs.

**Comment by fromanirh** [comment]
Let's have https://github.com/openshift-kni/debug-tools/pull/16 and https://github.com/openshift-kni/debug-tools/pull/17 before to move forward with this PR
Let's add tests (enabled by https://github.com/openshift-kni/debug-tools/pull/16) in `debug-tools` meanwhile.

**Comment by marcel-apf** [comment]
> actually we can do better than this with a trivial refactoring of knit, let me work on a couple PRs.

ack

File: must-gather/collection-scripts/gather_nodes [hunk 2]
```diff
@@ -64,7 +64,10 @@ do
     oc exec $pod -n perf-node-gather -- cat /proc/cmdline 2>/dev/null >> $NODE_PATH/proc_cmdline
     WORKER_NODES+=($node)
 
-    oc exec $pod -n perf-node-gather -- gather_sysinfo --debug --root=/host --output=- > $NODE_PATH/sysinfo.tgz 2> $NODE_PATH/sysinfo.log
+    oc exec $pod -n perf-node-gather -- gather_sysinfo --json cpuaff --procfs=/host/proc --sysfs=/host/sys > $NODE_PATH/cpu_affinities.json
+    oc exec $pod -n perf-node-gather -- gather_sysinfo --json irqaff --procfs=/host/proc --sysfs=/host/sys > $NODE_PATH/irq_affinities.json
```
Comment (fromanirh) [inline must-gather/collection-scripts/gather_nodes:68]: this step we can also do offline

File: must-gather/collection-scripts/gather_nodes [hunk 3]
```diff
@@ -64,7 +64,10 @@ do
     oc exec $pod -n perf-node-gather -- cat /proc/cmdline 2>/dev/null >> $NODE_PATH/proc_cmdline
     WORKER_NODES+=($node)
 
-    oc exec $pod -n perf-node-gather -- gather_sysinfo --debug --root=/host --output=- > $NODE_PATH/sysinfo.tgz 2> $NODE_PATH/sysinfo.log
+    oc exec $pod -n perf-node-gather -- gather_sysinfo --json cpuaff --procfs=/host/proc --sysfs=/host/sys > $NODE_PATH/cpu_affinities.json
```
Comment (fromanirh) [inline must-gather/collection-scripts/gather_nodes:67]: this step we cannot (yet) do it offline

**Comment by fromanirh** [comment]
/hold
let's discuss what we want to collect online and what not

File: must-gather/collection-scripts/gather_nodes [hunk 4]
```diff
@@ -66,6 +66,7 @@ do
 
     oc exec $pod -n perf-node-gather -- gather_sysinfo --json cpuaff --procfs=/host/proc --sysfs=/host/sys > $NODE_PATH/cpu_affinities.json
     oc exec $pod -n perf-node-gather -- gather_sysinfo --json irqaff --procfs=/host/proc --sysfs=/host/sys > $NODE_PATH/irq_affinities.json
+    oc exec $pod -n perf-node-gather -- gather_sysinfo --json podres --socket-path=/host/podresources/kubelet.sock > $NODE_PATH/podresources.json
```
Comment (fromanirh) [inline must-gather/collection-scripts/gather_nodes:69]: and this totally cannot do offline :)

File: must-gather/collection-scripts/gather_nodes (hunk 3 above)
Comment (fromanirh) [reply must-gather/collection-scripts/gather_nodes:67]: to do it offline we need https://github.com/openshift-kni/debug-tools/pull/20

File: must-gather/collection-scripts/gather_nodes (hunk 2 above)
Comment (fromanirh) [reply must-gather/collection-scripts/gather_nodes:68]: this data is already collected, we miss the user-friendly name but we have everything else.

Comment (marcel-apf) [inline must-gather/collection-scripts/gather_nodes:68]: it looks a little "off" we need to call the same tool twice and only change the irqaff / cpuaff, but is far from being an issue

File: tools/gather-sysinfo/gather-sysinfo.go [hunk 5]
```diff
@@ -1,79 +1,108 @@
...
+
 func main() {
-	debug := flag.Bool("debug", false, "enable debug output")
-	output := flag.String("output", "", "path to clone system information into")
-	rootDir := flag.String("root", "", "pseudofs root - use this if running inside a container")
-	dumpList := flag.Bool("dump", false, "just dump the glob list of expected content and exit")
+	root := cmd.NewRootCommand(newSnapshotCommand)
+
+	if err := root.Execute(); err != nil {
+		fmt.Fprintf(os.Stderr, "%v\n", err)
```
Comment (marcel-apf) [inline tools/gather-sysinfo/gather-sysinfo.go:26]: Are you sure we need to print the err? I think cobra already prints the error if Execute fails and we will end up with the same error twice: At the beginning and after the help output, with little/no difference between them. Is that the desired behavior?

File: tools/gather-sysinfo/gather-sysinfo.go [hunk 6]
```diff
@@ -1,79 +1,108 @@
...
+}
+
+func makeSnapshot(cmd *cobra.Command, knitOpts *cmd.KnitOptions, opts *snapshotOptions, args []string) error {
 	// ghw can't handle duplicates in CopyFilesInto, the operation will fail.
 	// Hence we need to make sure we just don't feed duplicates.
 	fileSpecs := dedupExpectedContent(kniExpectedCloneContent(), snapshot.ExpectedCloneContent())
-	if *dumpList {
+	if opts.dumpList {
 		for _, fileSpec := range fileSpecs {
 			fmt.Printf("%s\n", fileSpec)
```
Comment (marcel-apf) [inline tools/gather-sysinfo/gather-sysinfo.go:56]: to stdout? on purpose?

File: must-gather/node-gather/daemonset.yaml [hunk 7]
```diff
@@ -39,8 +39,10 @@ spec:
         volumeMounts:
           - name: sys
             mountPath: /host/sys
+            readOnly: true
           - name: proc
             mountPath: /host/proc
+            readOnly: true
           - name: podres
             mountPath: /host/podresources
```
Comment (marcel-apf) [inline must-gather/node-gather/daemonset.yaml:51]: Do we write to podres?

**Review by marcel-apf — COMMENTED** [review]
Some questions

File: tools/gather-sysinfo/gather-sysinfo.go (hunk 5 above)
Comment (fromanirh) [reply tools/gather-sysinfo/gather-sysinfo.go:26]: good point, will check

File: tools/gather-sysinfo/gather-sysinfo.go (hunk 6 above)
Comment (fromanirh) [reply tools/gather-sysinfo/gather-sysinfo.go:56]: yep. The list of the glob patterns becomes the expected output, hence we send it on stdout.

File: must-gather/node-gather/daemonset.yaml (hunk 7 above)
Comment (fromanirh) [reply must-gather/node-gather/daemonset.yaml:51]: yes: we need to write the request messages to send them to the server.

File: tools/gather-sysinfo/gather-sysinfo.go (hunk 5 above)
Comment (fromanirh) [reply tools/gather-sysinfo/gather-sysinfo.go:26]: current code:
```
$ ./build/_output/bin/gather-sysinfo irqaff -X
unknown shorthand flag: 'X' in -X
$ ./build/_output/bin/gather-sysinfo foo
unknown command "foo" for "knit"
$ ./build/_output/bin/gather-sysinfo
Usage:
  knit [flags]
  knit [command]

Available Commands:
  cpuaff      show cpu thread affinities
  help        Help about any command
  irqaff      show IRQ/softirq thread affinities
  podres      show currently allocated pod resources
  snapshot    snapshot pseudofilesystems for offline analysis
  wait        wait forever, or until a UNIX signal (SIGINT, SIGTERM) arrives

Flags:
  -C, --cpulist string   isolated cpu set to check (see man (7) cpuset - List format (default "0-16383")
  -D, --debug            enable debug log
  -h, --help             help for knit
  -J, --json             output as JSON
  -P, --procfs string    procfs root (default "/proc")
  -S, --sysfs string     sysfs root (default "/sys")

Use "knit [command] --help" for more information about a command.
```
without the line above:
```
$ ./build/_output/bin/gather-sysinfo irqaff -X
1060 13:20:33 fromani@musashi2 ~/Projects/golang/src/github.com/openshift-kni/performance-addon-operators $ echo $?
1
$ ./build/_output/bin/gather-sysinfo foo
$ echo $?
1
$ ./build/_output/bin/gather-sysinfo
Usage:
  knit [flags]
  knit [command]

Available Commands:
  cpuaff      show cpu thread affinities
  help        Help about any command
  irqaff      show IRQ/softirq thread affinities
  podres      show currently allocated pod resources
  snapshot    snapshot pseudofilesystems for offline analysis
  wait        wait forever, or until a UNIX signal (SIGINT, SIGTERM) arrives

Flags:
  -C, --cpulist string   isolated cpu set to check (see man (7) cpuset - List format (default "0-16383")
  -D, --debug            enable debug log
  -h, --help             help for knit
  -J, --json             output as JSON
  -P, --procfs string    procfs root (default "/proc")
  -S, --sysfs string     sysfs root (default "/sys")

Use "knit [command] --help" for more information about a command.
```

Comment (fromanirh) [reply tools/gather-sysinfo/gather-sysinfo.go:26]: So I'd say let's keep it.

File: must-gather/collection-scripts/gather_nodes (hunk 2 above)
Comment (fromanirh) [reply must-gather/collection-scripts/gather_nodes:68]: ok, keeping this for now

File: tools/gather-sysinfo/gather-sysinfo.go (hunk 5 above)
Comment (marcel-apf) [reply tools/gather-sysinfo/gather-sysinfo.go:26]: for sure... thanks. Weird, but we have  a different behavior in PPC.
/lgtm

**Pull request merged**